applications: added new endpoint(smart-plug) to matter-bridge (#16498)
nrfconnect/sdk-nrf: applications/matter_bridge/doc/matter_bridge_description.rst

.. note::
    Currently the DFU over Bluetooth LE SMP and Matter OTA are not available for the ``nrf54h20dk/nrf54h20/cpuapp`` board target.

> There is an additional functionality, which allows the Matter bridge device to work as a smart plug device.

Reviewer: I would add one more sentence after that. I think we could inform that in this scenario, the Matter bridge exposes additional endpoint with id ``2`` that represents Matter on/off smart plug device type functionality.

Thanks to it, you can integrate the Matter bridge functionality in a real end product, like smart plug, and do not have to make a bridge device work as a standalone device, which is much less attractive option.

> It is an optional feature, and can be enabled by passing accurate comman, see the :ref:`configuration_matter_bridge_functionality` section.

Reviewer: I would add one extra newline before this sentence. I will be better exposed on the webpage.

.. note::
    Currently the DFU over Bluetooth LE SMP and Matter OTA are not available for the ``nrf54h20dk/nrf54h20/cpuapp`` board target.

The Matter bridge device has an additional functionality, enabling it to work as a smart plug.

> This feature provides an additional endpoint with an ID equal to 2, which represents Matter on/off smart plug device type functionality.

Reviewer 2: ```suggestion
This feature provides an additional endpoint with an ID equal to 2.
The endpoint represents Matter On/Off Plug-in Unit :ref:`smart plug <ug_matter_device_types_plugs_outlets>` device type functionality.
```
I think that's correct.

.. _matter_bridge_app_bridged_support:

> There is an additional Functionality which allow matter bridge device to work as smart plug device.

Reviewer: ```suggestion
There is an additional functionality, which allows the Matter bridge device to work as a smart plug device.
```

> Thanks to it, we can integrate bridge functionality in a real device which is smart plug, and do not have create a bridge device as standalone device.

Reviewer: We don't use "we" form in documentation. It's "you" or impersonal form.
```suggestion
Thanks to it, you can integrate the Matter bridge functionality in a real end product, like smart plug, and do not have to make a bridge device work as a standalone device, which is much less attractive option.
```

> It is optional Feautre, and can be enabled by passing below building commands.

Reviewer: Below is where? :) I would suggest putting here a ref link to the configuring section.
```suggestion
It is an optional feature, and can be enabled by passing below building commands.
```

Reviewer: This comment should not be resolved as well. You write "below command", but below there is end of paragraph. I would suggest creating a ref build to the configuring section you have added and using it here as a link to direct user there.

         If pressed while the Matter smart plug functionality is enabled, the button changes the state of the smart plug device.

      LED 0:
         .. include:: /includes/matter_sample_state_led.txt

      LED 1:
         If the :ref:`CONFIG_BRIDGED_DEVICE_BT <CONFIG_BRIDGED_DEVICE_BT>` Kconfig option is set to ``y``, shows the current state of Bridge's Bluetooth LE connectivity.
         The following states are possible:

         * Turned Off - The Bridge device is in the idle state and has no Bluetooth LE devices paired.
         * Solid On - The Bridge device is in the idle state and all connections to the Bluetooth LE bridged devices are stable.
         * Slow Even Flashing (1000 ms on / 1000 ms off) - The Bridge device lost connection to at least one Bluetooth LE bridged device.
         * Even Flashing (300 ms on / 300 ms off) - The scan for Bluetooth LE devices is in progress.
         * Fast Even Flashing (100 ms on / 100 ms off) - The Bridge device is connecting to the Bluetooth LE device and waiting for the Bluetooth LE authentication PIN code.

         If used with the Matter smart plug functionality enabled, it shows the state of the smart plug device.

> .. include:: /includes/matter_segger_usb.txt

Reviewer 2: This include should be inside both tabs, for consistency.

> User interface

Reviewer 3: In the line 137 the exact button is mentioned, which if I understand correctly will be affected by the same issue with DK numbering. Can this part be generalized?
For example, ``If the Bluetooth LE advertising times out, you can re-enable it manually (see the :ref:`matter_bridge_app_ui` section). ``

> User interface

Reviewer 3: ```suggestion
.. _matter_bridge_app_ui:

User interface
```

Reviewer 3: The user interface section was unnecessarily removed, please add below the .. _matter_bridge_app_ui: part:
``User interface
**************``

**************

.. tabs::

>    .. group-tab:: nRF70 and nRF53 DKs

Reviewer 3: Just a nit here - in other samples we were usually numbering from the lowest to the highest number
```suggestion
   .. group-tab:: nRF53 and nRF70 DKs
```

Button 1:
   .. include:: /includes/matter_sample_button.txt

Button 2:

>    If we are using extend functionality, we can change the state of smart plug device by pressing it.

Reviewer: Avoid the "we" form, also extend functionality does not say much.
```suggestion
   In case the Matter smart plug functionality is used, you can change the state of smart plug device by pressing it.
```

       Enables only the necessary application functionalities to optimize its performance.

> Configure the functionality of the Matter-Bridge device

Reviewer: @annwoj  probably can give you more ideas of the documentation structure, but from my perspective I would suggest the following things:
- add the brief description telling about optional functionality that is enabling smart plug device type. It's worth to mention there that it present we can integrate bridge functionality in a real device which is smart plug and do not have to create a bridge device as a standalone device (such as router or Wi-Fi AP). I think it should be somewhere high on the page. Maybe as a subsection of `Overview` section.
- add the description how to enable this functionality using west command (what you've already done). I would suggest moving it into `Building and running` section as a separate subsection. Also you don;t have to describe how to build basic configuration, as it's already done in other place.
- update the `User interface` section to inform that LED2 and BUTTON2 are used in case of the smart plug variant.

Reviewer 3: For now I added some small suggestions, but agree with all the structure-related points above. Also, omitted the "basic functionality" part, as it seems it is covered

-------------------------------------------------------

The device can operate in a basic configuration and an extended one. The basic functionality means enabling communication on the matter-BLE line. Extended functionality, on the other hand, includes basic functionality and makes the device work as a smart-plug.

To use the basic functionality of the device, use the following command to build.

.. parsed-literal::
   :class: highlight

west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_WIFI_NRF700X=y -DCONFIG_CHIP_WIFI=y -Dmatter_bridge_SHIELD=nrf700x_nrf54h20dk -DCONFIG_BRIDGED_DEVICE_BT=y

> To use extened functionality, use following command.

Reviewer 3: ```suggestion
To enable the extended functionality, run the following command:
```

Before you start testing the application, you can select one of the :ref:`matter_bridge_app_custom_configs`.
See :ref:`app_build_file_suffixes` and :ref:`cmake_options` for more information how to select a configuration.

> .. _configuration_matter_bridge_functionality:

Reviewer: I'm not 100% if this is important or not, but usually all ref links within the same file starts with the same prefix, so I would recommend the following change. Also please remember to update the name above (in the place where it's used as a link).
```suggestion
.. _matter_bridge_smart_plug_functionality:
```

Configure the functionality of the Matter-Bridge device
-------------------------------------------------------

> To configure the Matter smart plug functionality, run the following command:

Reviewer: ```suggestion
To enable the Matter smart plug functionality, run the following command:
```

Before you start testing the application, you can select one of the :ref:`matter_bridge_app_custom_configs`.
See :ref:`app_build_file_suffixes` and :ref:`cmake_options` for more information how to select a configuration.

.. _matter_bridge_smart_plug_functionality:

Configure the functionality of the Matter-Bridge device
-------------------------------------------------------

To enable the Matter smart plug functionality, run the following command:

   ``nrf54h20dk/nrf54h20/cpuapp``

   .. code-block:: console

      west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_WIFI_NRF700X=y -DCONFIG_CHIP_WIFI=y -Dmatter_bridge_SHIELD=nrf700x_nrf54h20dk -DCONFIG_BRIDGED_DEVICE_BT=y -Dmatter_bridge_SNIPPET=onoff_plug

   ``nrf7002dk/nrf5340/cpuapp``

   .. code-block:: console

>       west build -b nrf7002dk/nrf5340/cpuapp -p -- -DCONFIG_BRIDGED_DEVICE_BT=y -Dmatter_bridge_SNIPPET=onoff_plug

Reviewer 3: Wouldn't it be better to move it into tabs?
And instead of using a board target as a title, to just say (similarly to what we did in samples):
nRF54 DKs   |  nRF(other) DKs, but listed, e.g., nRF52 and 53 DKs?

.. _matter_bridge_testing:

> Configure the functionality of the Matter-Bridge device

Reviewer: You cannot put it below .. _matter_bridge_testing: because this is a ref tag for the Testing section. It should be above it.

-------------------------------------------------------

> The device supports two configurations: basic and extended. 

Reviewer: I'm not sure whether we should describe basic configuration at all, as we have the commands for it in the Building and running section. Also I would recommend naming this "extended" configuration directly like "Configure the Matter smart plug functionality" or sth, as I don't think that using mystery extended name is necessary.

Reviewer: This was not resolved.

The basic functionality enables communication through Matter over Bluetooth LE. 
Extended functionality includes basic functionality and configures the device to work as a smart plug.

To use the basic functionality of the device, use the following command to build.

``nrf54h20dk/nrf54h20/cpuapp``

.. code-block:: console

   west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_WIFI_NRF700X=y -DCONFIG_CHIP_WIFI=y -Dmatter_bridge_SHIELD=nrf700x_nrf54h20dk -DCONFIG_BRIDGED_DEVICE_BT=y

``nrf7002dk/nrf5340/cpuapp``

.. code-block:: console

   west build -b nrf7002dk/nrf5340/cpuapp -p -- -DCONFIG_BRIDGED_DEVICE_BT=y

To enable the extended functionality, run the following command:

> .. code-block:: console

Reviewer: The  ``nrf54h20dk/nrf54h20/cpuapp`` is missing before this one. Also I'm not sure how graphically it's going to look on the page after rendering. If you modify the docs, you will always see the following comment in your PR that includes a link to the rendered docs, so you can check how it's going to look like:

![image](https://github.com/user-attachments/assets/18aca0bd-856a-4a0c-bdb2-6bf85e26cfe0)

Before you start testing the application, you can select one of the :ref:`matter_bridge_app_custom_configs`.
See :ref:`app_build_file_suffixes` and :ref:`cmake_options` for more information how to select a configuration.

.. _matter_bridge_smart_plug_functionality:

Configure the functionality of the Matter-Bridge device
-------------------------------------------------------

To enable the Matter smart plug functionality, run the following command:

.. tabs::

   .. group-tab:: nRF54 DKs

      .. code-block:: console

         west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_WIFI_NRF700X=y -DCONFIG_CHIP_WIFI=y -Dmatter_bridge_SHIELD=nrf700x_nrf54h20dk -DCONFIG_BRIDGED_DEVICE_BT=y -Dmatter_bridge_SNIPPET=onoff_plug

   .. group-tab:: nRF70 DKs

      .. code-block:: console

         west build -b nrf7002dk/nrf5340/cpuapp -p -- -DCONFIG_BRIDGED_DEVICE_BT=y -Dmatter_bridge_SNIPPET=onoff_plug

.. _matter_bridge_testing:

> Testing

Reviewer 3: I noticed that we also have the same DK numbering issue in the testing section. The solution we applied in other samples, to make it as easy as possible, was to create tabs, similarly to UI, and duplicating the entire testing sections with updated numbering. 
So I believe this solution would apply to "testing" and "Testing with bridged device working as a client" sections.
